src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.TreeNodes.cs

```diff
@@ -179,6 +247,59 @@ public IChecksumTreeNode GetOrCreateSubTreeNode<TKey>(TKey key)
                 return GetOrCreateChecksumObjectAsync(key, value, kind, valueGetterAsync, cancellationToken);
             }
 
+            protected void AppendChecksumObjects(
+                ImmutableDictionary<Checksum, ChecksumObject>.Builder builder,
+                HashSet<Checksum> checksums,
+                int itemCount,
+                IEnumerable<Checksum> itemChecksums,
+                Func<Checksum, ChecksumObject> itemGetter, CancellationToken cancellationToken)
+            {
+                using (var removed = Creator.CreateList<Checksum>())
+                {
+                    foreach (var checksum in GetChecksums(checksums.Count, checksums, itemCount, itemChecksums))
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        var checksumObject = itemGetter(checksum);
+                        if (checksumObject != null && checksums.Contains(checksum))
+                        {
+                            // found entry
+                            builder[checksum] = checksumObject;
+                            removed.Object.Add(checksum);
+
+                            if (removed.Object.Count == checksums.Count)
+                            {
+                                break;
+                            }
+                        }
+                    }
+
+                    checksums.ExceptWith(removed.Object);
+                }
+            }
+
+            private IEnumerable<Checksum> GetChecksums(int count1, IEnumerable<Checksum> checksums1, int count2, IEnumerable<Checksum> checksums2)
+            {
+                return count1 < count2 ? checksums1 : checksums2;
+            }
+
+            private void AppendChecksumObjects(
+                ImmutableDictionary<Checksum, ChecksumObject>.Builder builder, HashSet<Checksum> checksums, ChecksumObjectCache entry, CancellationToken cancellationToken)
+            {
+                AppendChecksumObjects(builder, checksums, entry.Count, entry.GetChecksums(), c => GetItem(entry, c), cancellationToken);
+            }
+
+            private ChecksumObject GetItem(ChecksumObjectCache entry, Checksum checksum)
+            {
+                ChecksumObject checksumObject;
+                if (entry.TryGetValue(checksum, out checksumObject))
+                {
+                    return checksumObject;
+                }
+
+                return null;
```

Comment:
yep, this subtree might not know about given checksum, but different sub tree might.